### PR TITLE
Test and fix for HHH-11437

### PR DESCRIPTION
**Pull request by beikov**
https://hibernate.atlassian.net/browse/HHH-11437

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/engine/internal/JoinSequence.java [hunk 1]
```diff
@@ -73,7 +73,7 @@ public JoinSequence getFromPart() {
 		return fromPart;
 	}
 
-	private Set<String> treatAsDeclarations;
+	protected Set<String> treatAsDeclarations;
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/JoinSequence.java]: Why are you making this protected?  I'd much prefer a protected getter if we need protected access

File: hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/EntityJoinFromElement.java [hunk 2]
```diff
@@ -129,10 +130,22 @@ public JoinFragment toJoinFragment(
 
 			final StringBuilder buffer = new StringBuilder();
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/EntityJoinFromElement.java]: Also, as we discussed on HipChat, please only apply this grouping when the inheritence is joined.

**Comment by Naros** [comment]
@beikov Nice work!  

This fix also addresses the issue with Envers [HHH-11416] where we have a subclass in a `JOINED` inheritance strategy which participates in an `INNER JOIN` with another entity and the columns of the subclass' super-class were not rendered in the SELECT clause, leading to missing column errors.  

I'll mark HHH-11416 as being fixed by HHH-11437.  Thanks!

**Comment by beikov** [comment]
I can squash together the changes which fix the issue if you want. Please note that the table group is only rendered when absolutely necessary i.e. in cases where currently wrong SQL is generated.

**Comment by dreab8** [comment]
applied upstream.

Thanks @beikov

**Pull request closed without merging**